### DISCUSSION TITLE
C++ stacktraces on exceptions

### DISCUSSION THREAD

**Discussion by arcanis** [emscripten-core/emscripten#17437] [Q&A] [answered]
I'm using the snippet described in this part of the documentation to catch C++ exceptions:

https://emscripten.org/docs/porting/Debugging.html#handling-c-exceptions-from-javascript

However, it doesn't print the stacktrace where the exception originally got thrown - is it possible to get this information?

**Comment by sbc100** [comment]
That example it designed to show how to get information on the thing that was thrown I believe.

For the stacktrace can you not just get it from `.stack` on the exception you catch?  Does devtools show the full stacktrace?

**Reply by arcanis** [reply]
There is no `Error` object which I could use to print a stacktrace - just a regular C++ exception, which doesn't keep stacktraces natively (at best I could create a new `Error` object inside the catch handler, but I would have the stacktrace for the catch, not the actual exception).

**Reply by sbc100** [reply]
Are you using wasm exceptions or emscripten exceptions?

@aheejin might know more.

**Reply by arcanis** [reply]
Emscripten exceptions. Basically I have a C++ `std::map<T>::at()` function throwing, but I can't know what part of the application threw it without bisecting with logs.

**Reply by sbc100** [reply]
When you catch exceptions in JS they should have the full stack trace attached.

For an example of this try running `./tests/runner core0.test_abort_on_exceptions -v`:

Here is part of the output:

```
-- begin program stderr --
Aborted(unhandled exception: Error: crash,Error: crash
    at throwException (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:1235:35)
    at cFunc (<anonymous>:wasm-function[16]:0x4b5)
    at /tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:993:23
    at /tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:972:22
    at /tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:3582:5
    at callRuntimeCallbacks (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:1432:26)
    at postRun (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:756:3)
    at doRun (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:3514:5)
    at run (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:3527:5)
    at runCaller (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:3425:19)
    at removeRunDependency (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:873:7)
    at receiveInstance (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:1134:5)
    at receiveInstantiationResult (/tmp/emscripten_test_core0_if8xjkp7/test_abort_on_exception.js:1152:5))
-- end program stderr --
ok
```

I tried modifying `tests/core/test_abort_on_exception_post.js` and I could see that each of the exceptions has a `.stack` attribute that works as expected.

**Reply by aheejin** [reply]
@sbc100 Is this about `abort`? 

I think Emscripten EH doesn't throw an `Error` instance, so stack traces are not supposed to be attached, no? https://github.com/emscripten-core/emscripten/blob/d2e80a6a4c16d3be8d7fd0ca69c4bacdfce70be0/src/parseTools.js#L687

I think it's worthwile to make an option to do the stack trace though, both of Emscripten EH and Wasm EH.

**Reply by sbc100** [reply]
No, not about abort.  Just using that test to show that the catch handler has access to the stack trace.

**Reply by sbc100** [reply]
Oh I see, the example I picked doesn't use C++ throw but throws an Error, which is why it gets a backtrace.  I agree having an option to get backtraces for C++ exceptions would be nice.

Also, would be nice to add a new setting by just tie it to existing `ASSERTIONS` settings

**Comment by aheejin** [comment]
I filed an issue for this (#17466) for the reminder to work on this, hopefully in near future. I'm not sure why it's not shown here. (Do discussions work differently from issues maybe?)

**Comment by aheejin** [comment, marked as answer]
#17979 will fix this for Wasm EH. Supporting stack traces in Emscripten EH doesn't seem to be easy, so we are not planning on that atm.

**Reply by major-mayer** [reply]
Great work!
Thanks a lot, this will make debugging a lot easier.
